### PR TITLE
Add --source_file_manifest flag to CoverageOutputGenerator.

### PR DESCRIPTION
**Pull request by iirina**
This new flag will allow C++ coverage collection to respect the `--instrumentation_filter` in the final coverage report.

### CODE REVIEW

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java [hunk 1]
```diff
@@ -80,6 +88,39 @@ public static void main(String[] args) {
     System.exit(exitStatus);
   }
 
+  /**
+   * Returns a set of source file names from the given manifest.
+   *
+   * The manifest contains file names line by line. Each file can either be a source file (e.g.
```
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: Done.

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java [hunk 2]
```diff
@@ -80,6 +88,39 @@ public static void main(String[] args) {
...
+   * .java, .cc) or a coverage metadata file (e.g. .gcno, .em).
+   *
+   * This method only returns the source files, ignoring the coverage metadata files as they are
+   * not relevant when putting together the final coverage report.
+   */
+  private static Set<String> getSourcesFromSourceFileManifest(String sourceFileManifest) {
+    Set<String> sourceFiles = new HashSet<>();
+    try (FileInputStream inputStream = new FileInputStream(new File(sourceFileManifest))) {
+      InputStreamReader inputStreamReader = new InputStreamReader(inputStream, UTF_8);
+      BufferedReader reader = new BufferedReader(inputStreamReader);
```
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: Thanks! TIL I can use multiple resources :)

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java [hunk 3]
```diff
@@ -80,6 +88,39 @@ public static void main(String[] args) {
...
+      for (String line = reader.readLine(); line != null; line = reader.readLine()) {
+        boolean isMetadataFile = false;
+        for (String metadataFileExt : COVERAGE_METADATA_FILE_EXTENSIONS) {
+          if (line.endsWith(metadataFileExt)) {
```
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: No, Bazel doesn't add whitespaces to the source file manifest.

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Coverage.java [hunk 4]
```diff
@@ -46,6 +47,32 @@ static Coverage merge(Coverage c1, Coverage c2) {
...
 
+  /**
+   * Returns {@link Coverage} only for the given source filenames, filtering out everything else in
+   * the given coverage away.
+   *
+   * @param coverage The initial coverage.
+   * @param sourcesToKeep The filenames of the sources to keep from the initial coverage.
+   */
+  static Coverage getOnlyTheseSources(Coverage coverage, Set<String> sourcesToKeep) {
+    if (coverage == null || sourcesToKeep == null) {
```
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Coverage.java]: No, if it's `null` then something went wrong.

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (laszlocsomor) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: Thanks! Sorry, I said it wrongly: start each paragraph **except the first one** with `<p>`.

[the remark this answers is not in the document]
Comment (laszlocsomor) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: (The JavaDoc documentation says to _separate_ paragraphs by `<p>` which is indeed an unambiguous way to say the same.)

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (laszlocsomor) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: Does `readLines()` trim the newline characters?

[the remark this answers is not in the document]
Comment (laszlocsomor) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: (I mean `read.readLine()`)

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: Oh sorry, done now.

File: tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (iirina) [reply tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main.java]: `readLine` reads everything until the next newline, so it does trim them

**Pull request closed without merging**